### ISSUE TITLE
wait() throws Error: Future must resolve before value is ready

### ISSUE THREAD

**Issue by veselov** [laverdet/node-fibers#273]
I don't have an SSCCE yet, but I'm seeing this after resolving a Future with a throw:

    at Object.Future.get (.../node_modules/fibers/future.js:210:10)
    at Object.Future.wait (.../node_modules/fibers/future.js:449:15)
 ...

On 1.0.10, node v4.4.0

I may be doing something funky, but I assume this should never get thrown when calling wait()...

**Comment by veselov** [comment]
Would this happen if I try to control a Fiber directly, and yet use Futures inside it?

**Comment by laverdet** [comment]
That could absolutely be the problem. What is your use case for that?

**Comment by veselov** [comment]
Yeah, I would really appreciate any help on this...
My base code has an assumption that there is a "current fiber", and uses futures left and right.
So, for any entry points from callbacks, I have to create a new fiber (if they already don't have one, which would be unlikely).
This becomes a problem for event emitters, for example for streams I noticed that "data" events end up being delivered after "end" event (which also probably means that the "data" events can get delivered in the wrong order).

**Comment by veselov** [comment]
OK, here is what I did.
The problem, really, is that anybody could "run" something on a Fiber, whether the Fiber was yielding because I was waiting for another chunk of "work", or because it was stopped on a future. So, I just need to distinguish between "my" yield, and future's yield.

            let f = new Fiber((target)=>{
                f.can_run = false;
                do {
                    try {
                        target();
                    } catch (e) {
                        // this just sends the error some place it belongs.
                        fail(e);
                    }   
                    if (f.queue.length) {
                        target = f.queue.shift();
                    } else {
                        f.can_run = true;
                        target = Fiber.yield();
                        f.can_run = false;
                    }   
                } while (target !== null);
            });         
                    
            f.queue = [];
            f.can_run = true;
  
            f.run0 = function(val) {
  
                if (f.can_run) {
                    f.run(val);
                } else {
                    f.queue.push(val);
                }

            };

Now, if I call `run0`, this will not interfere with any active future. Cool.

Thank you for Fibers :)

**Comment by laverdet** [comment]
Hey so the best way to handle event emitters or any other callback is either one of these two techniques:

```js
thing.on('data', function() {
  <you are now in a fiber>
}.future(true)); // <- this part is important
```

Or:
```js
thing.on('data', function() {
  Future.task(function() {
    <you are now in a fiber>
  }).detach();
}.future(true));
```

.future() will return a copy of your function which runs in a fiber and returns a future. Doing .future(true) will "detach" the future which means all exceptions thrown from that future will bubble up to the main node loop and eventually call process->uncaughtException which is important because you otherwise lose errors. The second way is identical but without using any convenience functions. Future.task() is cool because you can create multiple parallel fibers and run them all at once.

**Comment by veselov** [comment]
But if you use futures inside the event handler, wouldn't more events start showing up, and you would end up in a non-trivial situation? Also, the second variant in your suggestion doesn't need `future(true)`, does it?

May be I misunderstand what is conceptually different between these shortcuts and what I was doing, which was:

    thing.on('boo', () => { new Fiber( ()=> {
      // look, I'm inside fiber
    } ) ).run(); }

Once the function returned from `function(){}.future()` is called, or the `detach()` is called, the wrapped contents will be scheduled to run on some (new?) Fiber, and the current execution will continue (in this case to exit the event handler). And additional calls will to the handler will not re-use the same Fiber, right?

What seems to be happening in my case, is that when such an event handler is executed, it invokes Fiber.run() and somewhere there is a request for a Fiber to wake up/start is put on the exec queue. That request is racing with another invocation of another event handler. The event handler may come first, and put yet another Fiber request on the queue. These fiber requests may execute in "wrong" order. I know this sounds weird, I thought NodeJS guaranteed order of the queue, but I swear something is going on, and in some cases I do get stream's `end` before all of stream's `data`s. May be there is some library in the middle that's screwing this up, I can't really explain why that happens.

So I had to cook something up to make all event handlers to be processed serially.

**Comment by laverdet** [comment]
Yeh there was a typo in my second example from a copy/paste. You don't need the .future(true).

This is a pretty well-tested path of fibers, I'd be surprised if you found an actual issue here. I think you should get rid of all that fancy queue logic because I don't think it's helping you. The problem with interacting directly with fibers while also using futures is that all the context that futures works to maintain is lost. Take for instance:

```js
let current;
Fiber(function() {
  let future = new Future;
  process.nextTick(function() {
    future.return(1);
  });
  current = Fiber.current;
  future.wait();
}).run();
current.run(2);
```

This reproduces your same error-- "Error: Future must resolve before value is ready". If you're doing anything other than Fiber(...).run(), ie. manually calling `yield()` or `run()` on a running fiber you are very likely to hit issues like this. On the other hand if you only interact with Future it's very difficult to shoot yourself in the foot.

**Comment by veselov** [comment]
OK, I figured out how to show case the problem. And it's really a user logic problem, but it's a "reasonable" problem to run into.

Consider this:

    var handler = function(data) {
      saveToDB(data); // uses futures for c/b handling
      emitFurther(data); // sends to some next level of processing
    }

If I did `stream.on('data', handler.future(true));`, when the event is processed, the database processing time is undefined. As a result, there is no guarantee that `emitFurther` happens in the same sequence as emitting of the original data events.

The only solution (AFAIK), is to somehow synchronize the event handlers, and that's only possible (in Fibers world) if they are somehow tied to the same fiber. 

Unfortunately, since event handlers most likely come in from Node core, they won't have a Fiber. Methodologies like `Function.prototype.future`, `Future.type`, or creating and "running" a new Fiber - they all detach the event handler execution, making it impossible to make them run consecutively. 

Hence I tried using a shared fiber, where I ran into the future/fiber conflict. Which I could only solve by making sure that I call `run()` only when it's my `yield`, otherwise I schedule the next handler to run when the current one completes (as outlined in https://github.com/laverdet/node-fibers/issues/273#issuecomment-196619638).

**Comment by laverdet** [comment]
Why don't you just put `emitFurther` first, thus maintaining the order of your events.

You could also do something like this:
```js
var handler = function(data) {
  var dbSave = Future.task(function() {
    saveToDB(data); // uses futures for c/b handling
  });
  emitFurther(data); // sends to some next level of processing
  dbSave.wait();
}
```

Though it's worth noting in this case your `dbSave` task won't be called until the `.wait()` anyway.

If `emitFurther` for some reason depends on side-effects from `dbSave` you will need a more complex solution. The simplest approach would be to pause the stream while you're still processing the event, and resume again when your handler finishes.

I also generally recommend that you only call `wait()` from directly within a `Future.task()` or `.future()` function, that way you're never surprised when a function yields control unexpectedly. This is a personal choice though and doesn't have as much potential harm as directly calling `yield()`.

**Comment by veselov** [comment]
> Why don't you just put emitFurther first, thus maintaining the order of your events.

Because it's required that the data is passed to the next stage only if it was successfully stored in the database. So yes, in essence, it depends on the side-effects of the `dbSave`.

> The simplest approach would be to pause the stream while you're still processing the event, and resume again when your handler finishes.

That's a good suggestion, thank you. I do have some event sources that won't pause (though, I suppose, that can be implemented into them). It's worth noting that `pause()` must be called directly in the event handler, not in the wrapped function.
